jep/0000/README.adoc

```diff
@@ -0,0 +1,469 @@
+= JEP-0000: Evergreen Update Client/Server Lifecycle
+:toc: preamble
+:toclevels: 3
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+
+.Metadata
+[cols="2"]
+|===
+| JEP
+| 0000
+
+| Title
+| Evergreen Update Client/Server Lifecycle
+
+| Sponsor
+| link:https://github.com/rtyler[R. Tyler Croy]
+
+// Use the script `set-jep-status <jep-number> <status>` to update the status.
+| Status
+| Not Submitted :information_source:
+
+| Type
+| Standards
+
+| Created
+| 2018-04-17
+//
+//
+// Uncomment if there is an associated placeholder JIRA issue.
+//| JIRA
+//| :bulb: https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
+//
+//
+// Uncomment if there will be a BDFL delegate for this JEP.
+//| BDFL-Delegate
+//| :bulb: Link to github user page :bulb:
+//
+//
+// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
+//| Discussions-To
+//| :bulb: Link to where discussion and final status announcement will occur :bulb:
+//
+//
+// Uncomment if this JEP depends on one or more other JEPs.
+| Requires
+| JEP-300, JEP-303
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+
+== Abstract
+
+Integral to link:https://github.com/jenkinsci/jep/tree/master/jep/300[Jenkins
+Essentials] is the client's update lifecycle with the Evergreen hosted service
+layer. Starting with the first boot, the client must frequently inform the
+server of its current versions, retrieve new versions, and apply updates. This
+document specifies those API interactions and expectations to allow this update
+lifecycle to be performed correctly.
+
+== Specification
+
+There are two services involved behind the scenes which are responsible for
+facilitating the client's update lifecycle: `updates` and `versions`.
+
+The role of the versions service is simply to maintain an audit trail
+for the versions of software that each client has installed locally. As clients
+go through various upgrade processes, they will periodically `POST`
+<<version-manifest>> records which are stored in the Evergreen hosted service
+layer's storage.
```

Comment:
Do we _tolerate_ to miss version updates? Or should the client always push *all* versions/update levels it went through? I guess the latter.

Is the following use case potentially to happen?

* Client Pushes to versions service that it is on UL 9 (Update Level)
* Client upgrades to UL 10
* Client upgrades to UL 11
* What does the client push to `versions` service ? That it _went_ through 10, and 11. Or only "hey, I am on UL11" ?

Sorry if this is dumb question. Maybe it will clarified below. Reading.